### ISSUE TITLE
Validation runs only for one iteration when restarting from checkpoint mid-epoch, wrongly reporting validation loss

### ISSUE THREAD

**Issue by pimdh** [Lightning-AI/pytorch-lightning#19549]
### Bug description

When resuming from a mid-epoch checkpoint (which I have to use as my dataset is large), the training loop runs a validation loop for only one iteration, which leads to wrong validation loss logged.

It appears like the `batch_progress` of `lighting.pytorch.loops._EvaluationLoop` wrongly gets filled from the checkpoint as if the validation loop was already done, and not properly reset after the checkpoint is loaded.

### What version are you seeing the problem on?

v2.2

### How to reproduce the bug

```python
import os

import torch
from lightning.pytorch import LightningModule, Trainer
from lightning.pytorch.callbacks import ModelCheckpoint
from torch.utils.data import DataLoader, Dataset


class RandomDataset(Dataset):
    def __init__(self, size, num_samples):
        self.len = num_samples
        self.data = torch.randn(num_samples, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


training_iterations_to_go = 3


class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        global training_iterations_to_go

        if training_iterations_to_go == 0:
            raise KeyboardInterrupt()
        training_iterations_to_go -= 1
        loss = self(batch).sum()
        print("Training", batch.shape, loss.item(), batch_idx)
        self.log("train_loss", loss)
        return {"loss": loss}

    def validation_step(self, batch, batch_idx):
        loss = self(batch).sum()
        print("Validation", batch.shape, loss.item())
        self.log("valid_loss", loss)

    def test_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("test_loss", loss)

    def configure_optimizers(self):
        return torch.optim.SGD(self.layer.parameters(), lr=0.1)

    def on_train_epoch_end(self) -> None:
        return super().on_train_epoch_end()


def run(max_epochs, ckpt_path):
    train_data = DataLoader(RandomDataset(32, 64), batch_size=2)
    val_data = DataLoader(RandomDataset(32, 64), batch_size=2)

    model = BoringModel()
    trainer = Trainer(
        default_root_dir=os.getcwd(),
        limit_train_batches=8,
        limit_val_batches=3,
        val_check_interval=2,
        max_epochs=max_epochs,
        enable_model_summary=False,
        enable_progress_bar=False,
        logger=False,
        callbacks=[
            ModelCheckpoint(
                save_last=1,
                every_n_train_steps=1,
                dirpath="./checkpoints",
                enable_version_counter=False,
            )
        ],
    )
    trainer.fit(
        model,
        train_dataloaders=train_data,
        val_dataloaders=val_data,
        ckpt_path=ckpt_path,
    )


run(max_epochs=2, ckpt_path=None)
training_iterations_to_go = 3
run(max_epochs=4, ckpt_path="checkpoints/last.ckpt")
```

### Error messages and logs

```
GPU available: True (cuda), used: True
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
HPU available: False, using: 0 HPUs
/usr/local/lib/python3.10/dist-packages/lightning/pytorch/callbacks/model_checkpoint.py:652: Checkpoint directory /local/mnt/workspace/pim/projects/equi-scaling/checkpoints exists and is not empty.
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
/usr/local/lib/python3.10/dist-packages/lightning/pytorch/trainer/connectors/data_connector.py:441: The 'val_dataloader' does not have many workers which may be a bottleneck. Consider increasing the value of the `num_workers` argument` to `num_workers=7` in the `DataLoader` to improve performance.
Validation torch.Size([2, 32]) 0.07887113094329834
Validation torch.Size([2, 32]) -2.002589464187622
/usr/local/lib/python3.10/dist-packages/lightning/pytorch/trainer/connectors/data_connector.py:441: The 'train_dataloader' does not have many workers which may be a bottleneck. Consider increasing the value of the `num_workers` argument` to `num_workers=7` in the `DataLoader` to improve performance.
Training torch.Size([2, 32]) 0.5961282849311829 0
Training torch.Size([2, 32]) -1.8955397605895996 1
Validation torch.Size([2, 32]) -2.0098037719726562
Validation torch.Size([2, 32]) -0.34518975019454956
Validation torch.Size([2, 32]) -0.3607126474380493
Training torch.Size([2, 32]) 3.0296287536621094 2
/usr/local/lib/python3.10/dist-packages/lightning/pytorch/trainer/call.py:54: Detected KeyboardInterrupt, attempting graceful shutdown...
GPU available: True (cuda), used: True
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
HPU available: False, using: 0 HPUs
Restoring states from the checkpoint path at checkpoints/last.ckpt
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
Restored all states from the checkpoint at checkpoints/last.ckpt
/usr/local/lib/python3.10/dist-packages/lightning/pytorch/loops/training_epoch_loop.py:156: You're resuming from a checkpoint that ended before the epoch ended. This can cause unreliable results if further training is done. Consider using an end-of-epoch checkpoint
Validation torch.Size([2, 32]) -1.7108062505722046      # <----------------------- validation runs only for one iteration
Training torch.Size([2, 32]) -8.434117317199707 2
Training torch.Size([2, 32]) 3.0585873126983643 3
Validation torch.Size([2, 32]) -7.363028049468994
Validation torch.Size([2, 32]) -7.216525554656982
Validation torch.Size([2, 32]) -5.194650650024414
Training torch.Size([2, 32]) -7.072639465332031 4
```

### Environment

<details>
  <summary>Current environment</summary>
```
* CUDA:
        - GPU:
                - NVIDIA GeForce RTX 2080 Ti
        - available:         True
        - version:           11.7
* Lightning:
        - lightning:         2.2.0.post0
        - lightning-utilities: 0.10.1
        - pytorch-lightning: 2.2.0.post0
        - torch:             2.0.1
        - torch-ema:         0.3
        - torch-geometric:   2.5.0
        - torch-scatter:     2.1.2+pt20cu117
        - torchmetrics:      1.3.1
        - torchvision:       0.15.2
* System:
        - OS:                Linux
        - architecture:
                - 64bit
        - processor:         x86_64
        - python:            3.10.13
        - release:           5.4.0-152-generic
        - version:           #169-Ubuntu SMP Tue Jun 6 22:23:09 UTC 2023
```
</details>

### More info

Thanks a lot!

**Comment by awaelchli** [comment]
@pimdh Thank you for already investigating this.

Since the training loop is quite complex, I can't say for sure this is the right solution but it sounds reasonable. Would you be interested to send a PR with this change? We can then let the full test suite run on your PR and see if there are any edge cases with this. If it works I can help add a test case.

**Comment by pimdh** [comment]
Hi @awaelchli , I've filed the PR at #19583.
Thanks

**Comment by bkntr** [comment]
The same bug also happens when resuming training after being trained on IterableDataset
